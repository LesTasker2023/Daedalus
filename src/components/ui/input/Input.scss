// Input Titan Component - Enhanced SCSS
@use "../../../styles/tokens" as *;

.tk-input-container {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
  width: 100%;
}

.tk-input-wrapper {
  position: relative;
  display: flex;
  align-items: center;
  width: 100%;
}

.tk-input {
  // Base styles
  display: flex;
  width: 100%;
  border-radius: 0.375rem;
  border: 1px solid var(--color-border);
  background-color: var(--color-background);
  padding: 0.5rem 0.75rem;
  font-size: 0.875rem;
  line-height: 1.25rem;
  color: var(--color-foreground);
  transition: all 0.2s ease-in-out;
  box-shadow: 0 1px 2px 0 rgba(0, 0, 0, 0.05);

  &::placeholder {
    color: var(--color-muted-foreground);
  }

  &:focus {
    outline: none;
    border-color: var(--color-ring);
    box-shadow:
      0 0 0 2px var(--color-ring),
      0 4px 6px -1px rgba(0, 0, 0, 0.1),
      0 2px 4px -1px rgba(0, 0, 0, 0.06);
    transform: translateY(-1px);
  }

  &:disabled {
    cursor: not-allowed;
    opacity: 0.5;
    background-color: var(--color-muted);
  }

  // File input styles
  &[type="file"] {
    &::file-selector-button {
      border: 0;
      background-color: transparent;
      font-size: 0.875rem;
      font-weight: 500;
      color: var(--color-foreground);
      margin-right: 0.75rem;
      cursor: pointer;
    }
  }

  // Variant styles
  &--default {
    border-color: var(--color-border);
    background-color: var(--color-background);

    &:hover:not(:disabled) {
      border-color: var(--color-border-hover);
      box-shadow:
        0 4px 6px -1px rgba(0, 0, 0, 0.1),
        0 2px 4px -1px rgba(0, 0, 0, 0.06);
    }
  }

  &--outline {
    border: 2px solid var(--color-border);
    background-color: transparent;

    &:hover:not(:disabled) {
      border-color: var(--color-primary);
      background-color: var(--color-background);
    }

    &:focus {
      border-color: var(--color-primary);
      background-color: var(--color-background);
    }
  }

  &--filled {
    border: 1px solid transparent;
    background-color: var(--color-muted);

    &:hover:not(:disabled) {
      background-color: var(--color-muted-hover);
    }

    &:focus {
      background-color: var(--color-background);
      border-color: var(--color-ring);
    }
  }

  &--ghost {
    border: 1px solid transparent;
    background-color: transparent;

    &:hover:not(:disabled) {
      background-color: var(--color-accent);
      border-color: var(--color-border);
    }

    &:focus {
      background-color: var(--color-background);
      border-color: var(--color-ring);
    }
  }

  // Size variants
  &--sm {
    height: 2rem;
    padding: 0.25rem 0.5rem;
    font-size: 0.75rem;
    line-height: 1rem;
    border-radius: 0.25rem;
  }

  &--default-size {
    height: 2.5rem;
    padding: 0.5rem 0.75rem;
    font-size: 0.875rem;
    line-height: 1.25rem;
  }

  &--lg {
    height: 3rem;
    padding: 0.75rem 1rem;
    font-size: 1rem;
    line-height: 1.5rem;
    border-radius: 0.5rem;
  }

  // State variants
  &--error {
    border-color: var(--color-destructive);
    background-color: var(--color-destructive-background);

    &:focus {
      border-color: var(--color-destructive);
      box-shadow:
        0 0 0 2px var(--color-destructive),
        0 4px 6px -1px rgba(239, 68, 68, 0.1);
    }
  }

  &--success {
    border-color: var(--color-success);
    background-color: var(--color-success-background);

    &:focus {
      border-color: var(--color-success);
      box-shadow:
        0 0 0 2px var(--color-success),
        0 4px 6px -1px rgba(34, 197, 94, 0.1);
    }
  }

  &--warning {
    border-color: var(--color-warning);
    background-color: var(--color-warning-background);

    &:focus {
      border-color: var(--color-warning);
      box-shadow:
        0 0 0 2px var(--color-warning),
        0 4px 6px -1px rgba(245, 158, 11, 0.1);
    }
  }

  // Icon spacing
  .tk-input-wrapper:has(.tk-input__left-icon) & {
    padding-left: 2.5rem;
  }

  .tk-input-wrapper:has(.tk-input__right-icon) & {
    padding-right: 2.5rem;
  }
}

.tk-input__label {
  font-size: 0.875rem;
  font-weight: 500;
  color: var(--color-foreground);
  line-height: 1.25rem;
}

.tk-input__left-icon,
.tk-input__right-icon {
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  display: flex;
  align-items: center;
  justify-content: center;
  color: var(--color-muted-foreground);
  pointer-events: none;
  z-index: 1;
}

.tk-input__left-icon {
  left: 0.75rem;
}

.tk-input__right-icon {
  right: 0.75rem;
}

.tk-input__message {
  font-size: 0.75rem;
  line-height: 1rem;
  color: var(--color-muted-foreground);

  &--error {
    color: var(--color-destructive);
  }

  &--success {
    color: var(--color-success);
  }
}

// Responsive adjustments
@media (min-width: 768px) {
  .tk-input {
    font-size: 0.875rem;
  }

  .tk-input--lg {
    font-size: 1rem;
  }
}

// Dark mode enhancements
@media (prefers-color-scheme: dark) {
  .tk-input {
    box-shadow:
      0 1px 2px 0 rgba(0, 0, 0, 0.3),
      inset 0 1px 0 0 rgba(255, 255, 255, 0.05);

    &:focus {
      box-shadow:
        0 0 0 2px var(--color-ring),
        0 4px 6px -1px rgba(0, 0, 0, 0.3),
        0 2px 4px -1px rgba(0, 0, 0, 0.2),
        inset 0 1px 0 0 rgba(255, 255, 255, 0.1);
    }

    &:hover:not(:disabled) {
      box-shadow:
        0 4px 6px -1px rgba(0, 0, 0, 0.3),
        0 2px 4px -1px rgba(0, 0, 0, 0.2),
        inset 0 1px 0 0 rgba(255, 255, 255, 0.05);
    }
  }
}

// High contrast mode
@media (prefers-contrast: high) {
  .tk-input {
    border-width: 2px;

    &:focus {
      border-width: 3px;
    }
  }
}

// Reduced motion
@media (prefers-reduced-motion: reduce) {
  .tk-input {
    transition: none;
  }
}
