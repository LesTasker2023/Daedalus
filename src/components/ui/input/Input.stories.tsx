import type { Meta, StoryObj } from '@storybook/nextjs';
import { Input } from './Input';

const meta: Meta<typeof Input> = {
  title: 'UI/Input',
  component: Input,
  parameters: {
    layout: 'centered',
    docs: {
      description: {
        component:
          'A versatile input component with multiple variants, sizes, states, and enhanced accessibility features.',
      },
    },
  },
  tags: ['autodocs'],
  argTypes: {
    variant: {
      control: 'select',
      options: ['default', 'outline', 'filled', 'ghost'],
      description: 'The visual style variant of the input',
    },
    size: {
      control: 'select',
      options: ['sm', 'default', 'lg'],
      description: 'The size of the input',
    },
    state: {
      control: 'select',
      options: ['default', 'error', 'success', 'warning'],
      description: 'The validation state of the input',
    },
    asChild: {
      control: 'boolean',
      description: 'When true, the component will render as a Slot',
    },
    disabled: {
      control: 'boolean',
      description: 'Whether the input is disabled',
    },
    label: {
      control: 'text',
      description: 'Label text for the input',
    },
    helperText: {
      control: 'text',
      description: 'Helper text displayed below the input',
    },
    errorMessage: {
      control: 'text',
      description: 'Error message (overrides helperText and sets error state)',
    },
    successMessage: {
      control: 'text',
      description: 'Success message (overrides helperText and sets success state)',
    },
    placeholder: {
      control: 'text',
      description: 'Placeholder text',
    },
    type: {
      control: 'select',
      options: ['text', 'email', 'password', 'number', 'tel', 'url', 'search'],
      description: 'HTML input type',
    },
  },
};

export default meta;
type Story = StoryObj<typeof meta>;

// Basic variants
export const Default: Story = {
  args: {
    placeholder: 'Enter text...',
  },
};

export const Outline: Story = {
  args: {
    variant: 'outline',
    placeholder: 'Enter text...',
  },
};

export const Filled: Story = {
  args: {
    variant: 'filled',
    placeholder: 'Enter text...',
  },
};

export const Ghost: Story = {
  args: {
    variant: 'ghost',
    placeholder: 'Enter text...',
  },
};

// Sizes
export const Small: Story = {
  args: {
    size: 'sm',
    placeholder: 'Small input...',
  },
};

export const Large: Story = {
  args: {
    size: 'lg',
    placeholder: 'Large input...',
  },
};

// States
export const Error: Story = {
  args: {
    state: 'error',
    placeholder: 'Enter text...',
    defaultValue: 'invalid@email',
  },
};

export const Success: Story = {
  args: {
    state: 'success',
    placeholder: 'Enter text...',
    defaultValue: 'valid@email.com',
  },
};

export const Warning: Story = {
  args: {
    state: 'warning',
    placeholder: 'Enter text...',
    defaultValue: 'warning@email.com',
  },
};

// With labels and messages
export const WithLabel: Story = {
  args: {
    label: 'Email Address',
    placeholder: 'Enter your email...',
  },
};

export const WithHelperText: Story = {
  args: {
    label: 'Username',
    helperText: 'Must be at least 3 characters long',
    placeholder: 'Enter username...',
  },
};

export const WithErrorMessage: Story = {
  args: {
    label: 'Email',
    errorMessage: 'Please enter a valid email address',
    placeholder: 'Enter email...',
    defaultValue: 'invalid-email',
  },
};

export const WithSuccessMessage: Story = {
  args: {
    label: 'Email',
    successMessage: 'Email address is valid',
    placeholder: 'Enter email...',
    defaultValue: 'user@example.com',
  },
};

// Input types
export const EmailInput: Story = {
  args: {
    type: 'email',
    label: 'Email',
    placeholder: 'user@example.com',
  },
};

export const PasswordInput: Story = {
  args: {
    type: 'password',
    label: 'Password',
    placeholder: 'Enter password...',
  },
};

export const NumberInput: Story = {
  args: {
    type: 'number',
    label: 'Age',
    placeholder: '25',
  },
};

export const SearchInput: Story = {
  args: {
    type: 'search',
    placeholder: 'Search...',
    variant: 'filled',
  },
};

// With icons
export const WithLeftIcon: Story = {
  args: {
    label: 'Search',
    placeholder: 'Search...',
    leftIcon: '🔍',
  },
};

export const WithRightIcon: Story = {
  args: {
    label: 'Status',
    placeholder: 'Enter status...',
    rightIcon: '✓',
    state: 'success',
  },
};

export const WithBothIcons: Story = {
  args: {
    label: 'Search with status',
    placeholder: 'Search...',
    leftIcon: '🔍',
    rightIcon: '✓',
  },
};

// States
export const Disabled: Story = {
  args: {
    label: 'Disabled Input',
    placeholder: 'Cannot edit...',
    disabled: true,
    defaultValue: 'Disabled value',
  },
};

export const Required: Story = {
  args: {
    label: 'Required Field',
    placeholder: 'This field is required...',
    required: true,
    helperText: 'This field is required',
  },
};

// Complex example
export const ComplexForm: Story = {
  render: () => (
    <div className="space-y-4 w-96">
      <Input label="Full Name" placeholder="John Doe" required />
      <Input
        type="email"
        label="Email Address"
        placeholder="john@example.com"
        helperText="We'll never share your email"
        required
      />
      <Input
        type="password"
        label="Password"
        placeholder="Enter password"
        helperText="At least 8 characters"
        required
      />
      <Input type="tel" label="Phone Number" placeholder="+1 (555) 123-4567" variant="outline" />
      <Input label="Website" type="url" placeholder="https://example.com" variant="filled" />
    </div>
  ),
};

// Variant combinations
export const VariantShowcase: Story = {
  render: () => (
    <div className="space-y-4 w-96">
      <Input variant="default" placeholder="Default variant" />
      <Input variant="outline" placeholder="Outline variant" />
      <Input variant="filled" placeholder="Filled variant" />
      <Input variant="ghost" placeholder="Ghost variant" />
    </div>
  ),
};

export const SizeShowcase: Story = {
  render: () => (
    <div className="space-y-4 w-96">
      <Input size="sm" placeholder="Small size" />
      <Input size="default" placeholder="Default size" />
      <Input size="lg" placeholder="Large size" />
    </div>
  ),
};

export const StateShowcase: Story = {
  render: () => (
    <div className="space-y-4 w-96">
      <Input placeholder="Default state" />
      <Input state="error" placeholder="Error state" defaultValue="invalid input" />
      <Input state="success" placeholder="Success state" defaultValue="valid input" />
      <Input state="warning" placeholder="Warning state" defaultValue="warning input" />
    </div>
  ),
};
