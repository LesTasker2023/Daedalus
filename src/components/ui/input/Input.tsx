import { cn } from '@/lib/utils';
import { Slot } from '@radix-ui/react-slot';
import { cva, type VariantProps } from 'class-variance-authority';
import * as React from 'react';
import './Input.scss';

const inputVariants = cva('tk-input', {
  variants: {
    variant: {
      default: 'tk-input--default',
      outline: 'tk-input--outline',
      filled: 'tk-input--filled',
      ghost: 'tk-input--ghost',
    },
    size: {
      sm: 'tk-input--sm',
      default: 'tk-input--default-size',
      lg: 'tk-input--lg',
    },
    state: {
      default: '',
      error: 'tk-input--error',
      success: 'tk-input--success',
      warning: 'tk-input--warning',
    },
  },
  defaultVariants: {
    variant: 'default',
    size: 'default',
    state: 'default',
  },
});

export interface InputProps
  extends Omit<React.ComponentProps<'input'>, 'size'>,
    VariantProps<typeof inputVariants> {
  asChild?: boolean;
  label?: string;
  helperText?: string;
  errorMessage?: string;
  successMessage?: string;
  leftIcon?: React.ReactNode;
  rightIcon?: React.ReactNode;
}

const Input = React.forwardRef<HTMLInputElement, InputProps>(
  (
    {
      className,
      variant,
      size,
      state,
      asChild = false,
      label,
      helperText,
      errorMessage,
      successMessage,
      leftIcon,
      rightIcon,
      type = 'text',
      ...props
    },
    ref
  ) => {
    const Comp = asChild ? Slot : 'input';

    // Determine current state based on props
    const currentState = errorMessage ? 'error' : successMessage ? 'success' : state;
    const statusMessage = errorMessage || successMessage || helperText;

    const inputElement = (
      <div className="tk-input-wrapper">
        {leftIcon && <div className="tk-input__left-icon">{leftIcon}</div>}
        <Comp
          type={type}
          className={cn(inputVariants({ variant, size, state: currentState }), className)}
          ref={ref}
          {...props}
        />
        {rightIcon && <div className="tk-input__right-icon">{rightIcon}</div>}
      </div>
    );

    if (label || statusMessage) {
      return (
        <div className="tk-input-container">
          {label && (
            <label className="tk-input__label" htmlFor={props.id}>
              {label}
            </label>
          )}
          {inputElement}
          {statusMessage && (
            <div
              className={cn('tk-input__message', {
                'tk-input__message--error': errorMessage,
                'tk-input__message--success': successMessage,
              })}
            >
              {statusMessage}
            </div>
          )}
        </div>
      );
    }

    return inputElement;
  }
);

Input.displayName = 'Input';

export { Input, inputVariants };
