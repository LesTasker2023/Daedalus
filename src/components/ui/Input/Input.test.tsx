import { render, screen } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import { Input } from './input';

describe('Input', () => {
  describe('Snapshots', () => {
    it('matches default snapshot', () => {
      const { container } = render(<Input placeholder="Default input" />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('Functionality', () => {
    it('renders correctly', () => {
      render(<Input data-testid="input" placeholder="Test input" />);
      expect(screen.getByTestId('input')).toBeInTheDocument();
    });

    it('handles disabled state', () => {
      render(<Input disabled data-testid="input" />);
      const input = screen.getByTestId('input');
      expect(input).toBeDisabled();
    });

    it('accepts placeholder', () => {
      render(<Input placeholder="Enter text" data-testid="input" />);
      const input = screen.getByTestId('input');
      expect(input).toHaveAttribute('placeholder', 'Enter text');
    });

    it('accepts different input types', () => {
      render(<Input type="email" data-testid="input" />);
      const input = screen.getByTestId('input');
      expect(input).toHaveAttribute('type', 'email');
    });
  });
});
