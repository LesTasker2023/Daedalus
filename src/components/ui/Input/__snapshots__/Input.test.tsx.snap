// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input > Snapshots > matches default snapshot 1`] = `
<input
  class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-base shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 md:text-sm input"
  data-testid="input"
/>
`;

exports[`Input > Snapshots > matches disabled state snapshot 1`] = `
<input
  class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-base shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 md:text-sm input"
  data-testid="input"
  disabled=""
/>
`;

exports[`Input > Snapshots > matches error state snapshot 1`] = `
<div
  class="input-wrapper"
>
  <input
    class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-base shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 md:text-sm input input--error"
    data-testid="input"
  />
  <p
    class="input-error text-sm text-destructive mt-1"
  />
</div>
`;

exports[`Input > Snapshots > matches loading state snapshot 1`] = `
<input
  class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-base shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 md:text-sm input input--loading"
  data-testid="input"
  disabled=""
/>
`;
